src/lib/uTop_main.ml

```diff
@@ -336,8 +336,14 @@ end = struct
   let scan_cmis =
     let new_cmis = ref [] in
     let default_load = !Persistent_env.Persistent_signature.load in
-    let load ~unit_name =
+
```

Comment:
I think that this part can be made into a `add_cmi_hook : (cmi_infos -> unit) -> unit` which will also improve `scan_cmis`.

Reply:
Where's `add_cmi_hook` coming from?

Reply:
Sorry, let me expand what I meant here.
`scan_cmis` does two things:
- it changes `Persistent_env.Persistent_signature.load` so that each time it's called, the result value is added to a `new_cmis` global (even if it has a local lexical scope)
- it returns a function that consumes this list

To support 5.2 we need to alter the first part. My suggestion is to cut the function in two parts:

- part 1 in `UTop_compat`
```ocaml
let add_cmi_hook f =
#if OCAML_VERSION...
  let load ...
#endif
    Option.iter (fun x -> f x.cmi) res
  in
  Persistent_env....load := load
```

- part 2 here:

```ocaml
let scan_cmis =
  let new_cmis = ref [] in
  add_cmi_hook (fun cmi -> new_cmis := cmi :: !new_cmis);
  fun pp ->
    ...
```